File: apps/server/src/indicators/indicators.ts
/**
 * 技术指标计算 — 精确翻译自 kline_indicators.py
 * 所有公式与 Python 版本完全一致，确保历史数据一致性。
 */

export interface KlineRow {
  open_time: Date | string;
  open: string | number;
  high: string | number;
  low: string | number;
  close: string | number;
  volume: string | number;
  close_time?: Date | string;
  quote_volume?: string | number;
  trades?: string | number;
  taker_buy_base_vol?: string | number;
  taker_buy_quote_vol?: string | number;
}

export interface KlineRowWithIndicators extends KlineRow {
  DIF: number;
  DEA: number;
  MACD: number;
  'KDJ.K': number;
  'KDJ.D': number;
  'KDJ.J': number;
  BBI: number;
  MA5: number;
  MA30: number;
  MA60: number;
  MA120: number;
  MA240: number;
  '10_quote_volume': number;
  atr_14: number;
  loss_atr_14: number;
  low_9: number;
  high_9: number;
  stop_loss_pct: number;
  risk_reward_ratio: number;
}

/** EMA — 首值以第一个数据为种子（Python: k = 2/(period+1)） */
function calcEma(values: number[], period: number): number[] {
  const k = 2.0 / (period + 1);
  const result: number[] = [];
  for (let i = 0; i < values.length; i++) {
    result.push(i === 0 ? values[i] : values[i] * k + result[i - 1] * (1 - k));
  }
  return result;
}

/** SMA — 不足 period 时取已有数据的均值 */
function calcSma(values: number[], period: number): number[] {
  const result: number[] = [];
  for (let i = 0; i < values.length; i++) {
    const start = Math.max(0, i - period + 1);
    const window = values.slice(start, i + 1);
    result.push(window.reduce((a, b) => a + b, 0) / window.length);
  }
  return result;
}

/** 按有效数字位数四舍五入（Python: _round_sig） */
function roundSig(x: number, sig = 8): number {
  if (x === 0 || !isFinite(x)) return x;
  const magnitude = Math.floor(Math.log10(Math.abs(x)));
  const factor = Math.pow(10, Math.max(sig - 1 - magnitude, 0));
  return Math.round(x * factor) / factor;
}

/** Wilder's ATR — Python: _calc_atr */
function calcAtr(highs: number[], lows: number[], closes: number[], period: number): number[] {
  const n = highs.length;
  const tr: number[] = [highs[0] - lows[0]];
  for (let i = 1; i < n; i++) {
    tr.push(Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1]),
    ));
  }

  const atr: number[] = [];
  for (let i = 0; i < n; i++) {
    if (i < period - 1) {
      atr.push(tr.slice(0, i + 1).reduce((a, b) => a + b, 0) / (i + 1));
    } else if (i === period - 1) {
      atr.push(tr.slice(0, period).reduce((a, b) => a + b, 0) / period);
    } else {
      atr.push((atr[i - 1] * (period - 1) + tr[i]) / period);
    }
  }
  return atr;
}

/**
 * 原地计算并填充技术指标
 * 精确翻译自 kline_indicators.py calc_indicators()
 */
export function calcIndicators(rows: KlineRow[]): KlineRowWithIndicators[] {
  const closes = rows.map((r) => parseFloat(String(r.close)));
  const highs = rows.map((r) => parseFloat(String(r.high)));
  const lows = rows.map((r) => parseFloat(String(r.low)));
  const n = closes.length;

  // MACD
  const ema12 = calcEma(closes, 12);
  const ema26 = calcEma(closes, 26);
  const dif = ema12.map((v, i) => v - ema26[i]);
  const dea = calcEma(dif, 9);
  const macd = dif.map((d, i) => 2.0 * (d - dea[i]));

  // KDJ（周期 9，初始 K=D=50）
  const kVals: number[] = [];
  const dVals: number[] = [];
  const jVals: number[] = [];
  let prevK = 50.0;
  let prevD = 50.0;
  for (let i = 0; i < n; i++) {
    const s = Math.max(0, i - 8);
    const hMax = Math.max(...highs.slice(s, i + 1));
    const lMin = Math.min(...lows.slice(s, i + 1));
    const rsv = hMax !== lMin
      ? ((closes[i] - lMin) / (hMax - lMin)) * 100
      : 50.0;
    const k = prevK * (2 / 3) + rsv / 3;
    const d = prevD * (2 / 3) + k / 3;
    kVals.push(k);
    dVals.push(d);
    jVals.push(3 * k - 2 * d);
    prevK = k;
    prevD = d;
  }

  // BBI = (MA3+MA6+MA12+MA24)/4
  const sma3 = calcSma(closes, 3);
  const sma6 = calcSma(closes, 6);
  const sma12 = calcSma(closes, 12);
  const sma24 = calcSma(closes, 24);
  const bbi = sma3.map((v, i) => (v + sma6[i] + sma12[i] + sma24[i]) / 4);

  const ma5 = calcSma(closes, 5);
  const ma30 = calcSma(closes, 30);
  const ma60 = calcSma(closes, 60);
  const ma120 = calcSma(closes, 120);
  const ma240 = calcSma(closes, 240);

  const qvols = rows.map((r) => parseFloat(String(r.quote_volume || 0)));
  const qvol10 = calcSma(qvols, 10);
  const atr14 = calcAtr(highs, lows, closes, 14);

  // 9日高低、止损幅度、盈亏比
  const low9: number[] = [];
  const high9: number[] = [];
  const slPct: number[] = [];
  const rr: number[] = [];
  for (let i = 0; i < n; i++) {
    const s = Math.max(0, i - 8);
    const h9 = Math.max(...highs.slice(s, i + 1));
    const l9 = Math.min(...lows.slice(s, i + 1));
    high9.push(h9);
    low9.push(l9);
    slPct.push(closes[i] ? (1 - l9 / closes[i]) * 100 : 0.0);
    const loss = closes[i] - l9;
    rr.push(loss ? (h9 - closes[i]) / loss : 0.0);
  }

  return rows.map((row, i) => ({
    ...row,
    DIF: roundSig(dif[i], 8),
    DEA: roundSig(dea[i], 8),
    MACD: roundSig(macd[i], 8),
    'KDJ.K': parseFloat(kVals[i].toFixed(4)),
    'KDJ.D': parseFloat(dVals[i].toFixed(4)),
    'KDJ.J': parseFloat(jVals[i].toFixed(4)),
    BBI: roundSig(bbi[i], 8),
    MA5: roundSig(ma5[i], 8),
    MA30: roundSig(ma30[i], 8),
    MA60: roundSig(ma60[i], 8),
    MA120: roundSig(ma120[i], 8),
    MA240: roundSig(ma240[i], 8),
    '10_quote_volume': parseFloat(qvol10[i].toFixed(2)),
    atr_14: roundSig(atr14[i], 8),
    loss_atr_14: roundSig(closes[i] - atr14[i], 8),
    low_9: roundSig(low9[i], 8),
    high_9: roundSig(high9[i], 8),
    stop_loss_pct: parseFloat(slPct[i].toFixed(4)),
    risk_reward_ratio: roundSig(rr[i], 4),
  })) as KlineRowWithIndicators[];
}

File: apps/quant-pipeline/src/quant_pipeline/custom_index/indicators.py
"""自定义指数技术指标（port NestJS indicators-stream / us_indicators）。"""

from __future__ import annotations

import math
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any

from quant_pipeline.custom_index.types import IndexQuoteRow
from quant_pipeline.sync.us_indicators import (
    _ema,
    _round_fixed,
    _round_sig,
    _sma,
    _strict_sma,
)

BRICK_P = 4
BRICK_N1 = 4
BRICK_N2 = 6


@dataclass
class _BrickState:
    sma2a: float = 0.0
    sma4a: float = 0.0
    sma5a: float = 0.0
    prev1: float = 0.0
    prev2: float = 0.0
    inited: bool = False
    highs: list[float] | None = None
    lows: list[float] | None = None


def _append_window(values: list[float], value: float, limit: int) -> list[float]:
    nxt = [*values, value]
    return nxt[-limit:] if len(nxt) > limit else nxt


def _calc_next_brick(
    prev: _BrickState | None,
    high: float,
    low: float,
    close: float,
) -> tuple[float, float, float, float]:
    highs = _append_window(prev.highs if prev else [], high, BRICK_P)
    lows = _append_window(prev.lows if prev else [], low, BRICK_P)
    hhv = max(highs)
    llv = min(lows)
    rng = hhv - llv
    var1a = (hhv - close) / rng * 100 - 90 if rng > 0 else -90.0
    var3a = (close - llv) / rng * 100 if rng > 0 else 50.0
    if prev is None or not prev.inited:
        sma2a, sma4a, sma5a = var1a, var3a, var3a
    else:
        sma2a = (var1a + (BRICK_N1 - 1) * prev.sma2a) / BRICK_N1
        sma4a = (var3a + (BRICK_N2 - 1) * prev.sma4a) / BRICK_N2
        sma5a = (sma4a + (BRICK_N2 - 1) * prev.sma5a) / BRICK_N2
    var6a = (sma5a + 100) - (sma2a + 100)
    brick = var6a - 4 if var6a > 4 else 0.0
    return brick, sma2a, sma4a, sma5a


def _calc_brick_delta(current: float, prev1: float, prev2: float) -> float:
    diff1 = abs(current - prev1)
    diff2 = abs(prev1 - prev2)
    return diff1 / diff2 if diff2 > 1e-10 else 0.0


def calc_index_indicators(quotes: list[IndexQuoteRow]) -> list[dict[str, Any]]:
    """从指数 close/high/low 序列计算 MA/MACD/KDJ/BBI/砖图。"""

    if not quotes:
        return []

    closes = [float(q.close) for q in quotes if q.close is not None]
    highs = [float(q.high) for q in quotes if q.high is not None]
    lows = [float(q.low) for q in quotes if q.low is not None]
    n = len(quotes)
    if n == 0 or len(closes) != n:
        return []

    ema12 = _ema(closes, 12)
    ema26 = _ema(closes, 26)
    dif = [ema12[i] - ema26[i] for i in range(n)]
    dea = _ema(dif, 9)
    macd = [2.0 * (dif[i] - dea[i]) for i in range(n)]

    k_vals: list[float] = []
    d_vals: list[float] = []
    j_vals: list[float] = []
    prev_k = 50.0
    prev_d = 50.0
    for i in range(n):
        s = max(0, i - 8)
        h_max = max(highs[s : i + 1])
        l_min = min(lows[s : i + 1])
        rsv = ((closes[i] - l_min) / (h_max - l_min)) * 100 if h_max != l_min else 50.0
        k = prev_k * (2 / 3) + rsv / 3
        d = prev_d * (2 / 3) + k / 3
        k_vals.append(k)
        d_vals.append(d)
        j_vals.append(3 * k - 2 * d)
        prev_k, prev_d = k, d

    sma3 = _sma(closes, 3)
    sma6 = _sma(closes, 6)
    sma12 = _sma(closes, 12)
    sma24 = _sma(closes, 24)
    bbi = [(sma3[i] + sma6[i] + sma12[i] + sma24[i]) / 4 for i in range(n)]

    ma5 = _strict_sma(closes, 5)
    ma30 = _strict_sma(closes, 30)
    ma60 = _strict_sma(closes, 60)
    ma120 = _strict_sma(closes, 120)
    ma240 = _strict_sma(closes, 240)

    brick_state: _BrickState | None = None
    out: list[dict[str, Any]] = []
    for i in range(n):
        brick, sma2a, sma4a, sma5a = _calc_next_brick(
            brick_state,
            highs[i],
            lows[i],
            closes[i],
        )
        prev1 = brick_state.prev1 if brick_state else 0.0
        prev2 = brick_state.prev2 if brick_state else 0.0
        aa = i >= 1 and brick > prev1
        aa_prev = i >= 2 and prev1 > prev2
        brick_delta = _calc_brick_delta(brick, prev1, prev2) if i >= 2 else 0.0
        brick_xg = i >= 2 and not aa_prev and aa

        prev_state = brick_state
        brick_state = _BrickState(
            sma2a=sma2a,
            sma4a=sma4a,
            sma5a=sma5a,
            prev1=brick,
            prev2=prev1,
            inited=True,
            highs=_append_window(prev_state.highs if prev_state else [], highs[i], BRICK_P),
            lows=_append_window(prev_state.lows if prev_state else [], lows[i], BRICK_P),
        )

        out.append(
            {
                "trade_date": quotes[i].trade_date,
                "ma5": _round_sig(ma5[i], 8),
                "ma30": _round_sig(ma30[i], 8),
                "ma60": _round_sig(ma60[i], 8),
                "ma120": _round_sig(ma120[i], 8),
                "ma240": _round_sig(ma240[i], 8),
                "dif": _round_sig(dif[i], 8),
                "dea": _round_sig(dea[i], 8),
                "macd": _round_sig(macd[i], 8),
                "kdj_k": _round_fixed(k_vals[i], 4),
                "kdj_d": _round_fixed(d_vals[i], 4),
                "kdj_j": _round_fixed(j_vals[i], 4),
                "bbi": _round_sig(bbi[i], 8),
                "brick": _round_sig(brick, 8),
                "brick_delta": _round_sig(brick_delta, 8),
                "brick_xg": brick_xg,
            }
        )
    return out


def calc_simple_ma(values: Sequence[float | None], period: int) -> list[float | None]:
    """AMV 均线辅助。"""

    out: list[float | None] = []
    buf: list[float] = []
    for v in values:
        if v is None or (isinstance(v, float) and math.isnan(v)):
            out.append(None)
            continue
        buf.append(float(v))
        window = buf[-period:]
        if len(window) < period:
            out.append(None)
        else:
            out.append(sum(window) / period)
    return out
